// @hermeznetwork/hermezjs/dist/node/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._setContractAddress = setContractAddress;
exports.BASE_API_URL = exports.contractAddresses = exports.ContractNames = exports.DEFAULT_PAGE_SIZE = exports.GAS_MULTIPLIER = exports.GAS_LIMIT = exports.ETHER_TOKEN_ID = exports.CREATE_ACCOUNT_AUTH_MESSAGE = exports.METAMASK_MESSAGE = exports.TRANSACTION_POOL_KEY = void 0;
var TRANSACTION_POOL_KEY = 'transactionPool';
exports.TRANSACTION_POOL_KEY = TRANSACTION_POOL_KEY;
var METAMASK_MESSAGE = 'HERMEZ_ACCOUNT. Don\'t share this signature with anyone as this would reveal your Hermez private key. Unless you are in a trusted application, DO NOT SIGN THIS';
exports.METAMASK_MESSAGE = METAMASK_MESSAGE;
var CREATE_ACCOUNT_AUTH_MESSAGE = 'I authorize this babyjubjub key for hermez rollup account creation';
exports.CREATE_ACCOUNT_AUTH_MESSAGE = CREATE_ACCOUNT_AUTH_MESSAGE;
var ETHER_TOKEN_ID = 0;
exports.ETHER_TOKEN_ID = ETHER_TOKEN_ID;
var GAS_LIMIT = 5000000;
exports.GAS_LIMIT = GAS_LIMIT;
var GAS_MULTIPLIER = 1;
exports.GAS_MULTIPLIER = GAS_MULTIPLIER;
var ContractNames = {
  Hermez: 'Hermez',
  WithdrawalDelayer: 'WithdrawalDelayer'
};
exports.ContractNames = ContractNames;
var contractAddresses = {
  Hermez: '0x10465b16615ae36F350268eb951d7B0187141D3B',
  WithdrawalDelayer: '0x8EEaea23686c319133a7cC110b840d1591d9AeE0'
};
/**
 * Tells hermezjs to use alternative contract addresses
 * Useful for testing with non-standardised addresses
 * @private
 */

exports.contractAddresses = contractAddresses;

function setContractAddress(name, address) {
  contractAddresses[ContractNames[name]] = address;
}

var DEFAULT_PAGE_SIZE = 20;
exports.DEFAULT_PAGE_SIZE = DEFAULT_PAGE_SIZE;
var BASE_API_URL = 'http://localhost:8086';
exports.BASE_API_URL = BASE_API_URL;
