// @hermeznetwork/hermezjs/dist/node/signers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSigner = exports.SignerType = void 0;

var _ethers = require("ethers");

var _ledgerSigner = require("./signers/ledger-signer.js");

var _trezorSigner = require("./signers/trezor-signer.js");

/**
 * Represents the different types of signers that can be built.
 */
var SignerType = {
  JSON_RPC: 'JSON-RPC',
  LEDGER: 'LEDGER',
  TREZOR: 'TREZOR',
  WALLET: 'WALLET'
};
/**
 * Builds an ethers Signer.
 * @param {Provider} provider - Ethers provider
 * @param {Object} signerData - Data required to build a signer
 */

exports.SignerType = SignerType;

var getSigner = function getSigner(provider, signerData) {
  if (!signerData) {
    return provider.getSigner();
  }

  switch (signerData.type) {
    case SignerType.LEDGER:
      {
        return _ledgerSigner.LedgerSigner.connect(provider, {
          path: signerData.path
        });
      }

    case SignerType.TREZOR:
      {
        return _trezorSigner.TrezorSigner.connect(provider, {
          manifest: signerData.manifest,
          path: signerData.path,
          address: signerData.address
        });
      }

    case SignerType.JSON_RPC:
      {
        return provider.getSigner(signerData.addressOrIndex);
      }

    case SignerType.WALLET:
      {
        return new _ethers.ethers.Wallet(signerData.privateKey, provider);
      }

    default:
      {
        return provider.getSigner();
      }
  }
};

exports.getSigner = getSigner;
