// @hermeznetwork/hermezjs/dist/node/signers/ledger-signer.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

require("core-js/modules/es.array.concat.js");

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.object.get-prototype-of.js");

require("core-js/modules/es.object.set-prototype-of.js");

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/es.regexp.to-string.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LedgerSigner = void 0;

var _hwTransportWebusb = _interopRequireDefault(require("@ledgerhq/hw-transport-webusb"));

var _hwAppEth = _interopRequireDefault(require("@ledgerhq/hw-app-eth"));

var _ethers = require("ethers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

/**
 * Ledger hardware wallet signer
 */
var LedgerSigner = /*#__PURE__*/function (_ethers$Signer) {
  _inherits(LedgerSigner, _ethers$Signer);

  var _super = _createSuper(LedgerSigner);

  /**
   * Creates a LedgerSigner instance
   * @param {Object} transport - Ledger transport
   * @param {Object} provider - Ethers provider
   * @param {Object} options - Additional data to set up the signer, e.g. hw wallet path
   */
  function LedgerSigner(transport, provider, options) {
    var _this;

    _classCallCheck(this, LedgerSigner);

    _this = _super.call(this);
    _this.eth = new _hwAppEth.default(transport);
    _this.provider = provider;
    _this.path = options && options.path ? options.path : _ethers.ethers.utils.defaultPath;
    return _this;
  }
  /**
   * Returns the checksum address
   * @returns {Promise} - Promise of the checksum address
   */


  _createClass(LedgerSigner, [{
    key: "getAddress",
    value: function getAddress() {
      return this.eth.getAddress(this.path).then(function (_ref) {
        var address = _ref.address;
        return address;
      });
    }
    /**
     * Returns the signed prefixed-message.
     * @param {Uint8[]|string} message - Message to sign in bytes[] or string
     * @returns {Promise} - Promise of the message signature
     */

  }, {
    key: "signMessage",
    value: function signMessage(message) {
      var messageBytes = typeof message === 'string' ? _ethers.ethers.utils.toUtf8Bytes(message) : message;

      var messageHex = _ethers.ethers.utils.hexlify(messageBytes);

      return this.eth.signPersonalMessage(this.path, messageHex.slice(2)).then(function (result) {
        var hexV = (result.v - 27).toString(16);
        var fixedHexV = hexV.length < 2 ? "0".concat(hexV) : hexV;
        var signature = "0x".concat(result.r).concat(result.s).concat(fixedHexV);
        return signature;
      });
    }
    /**
     * Signs a transaction
     * @param {Object} transaction - Transaction to be signed
     * @returns {Promise} - Promise of the transaction signature
     */

  }, {
    key: "signTransaction",
    value: function signTransaction(transaction) {
      var _this2 = this;

      return _ethers.ethers.utils.resolveProperties().then(function (tx) {
        var unsignedTx = _ethers.ethers.utils.serializeTransaction(tx).substr(2);

        return _this2.eth.signTransaction(_this2.path, unsignedTx).then(function (_ref2) {
          var r = _ref2.r,
              s = _ref2.s,
              v = _ref2.v;
          var signature = {
            r: "0x".concat(r),
            s: "0x".concat(s),
            v: v
          };
          return _ethers.ethers.utils.serializeTransaction(tx, signature);
        });
      });
    }
    /**
     * Returns a new instance of the Signer, connected to the provider.
     * @param {Object} provider - Ethers provider
     * @param {Object} options - Additional data to set up the signer, e.g. hw wallet path
     * @returns {Promise} - Promise of a new LedgerSigner instance
     */

  }], [{
    key: "connect",
    value: function connect(provider, options) {
      return _hwTransportWebusb.default.create().then(function (transport) {
        return new LedgerSigner(transport, provider, options);
      });
    }
  }]);

  return LedgerSigner;
}(_ethers.ethers.Signer);

exports.LedgerSigner = LedgerSigner;
